src/ol_infrastructure/components/services/vault.py

```diff
@@ -22,47 +22,58 @@
 DEFAULT_PORT_POSTGRES = 5432
 DEFAULT_PORT_MYSQL = 3306
 DEFAULT_POT_MONGODB = 27017
-SIX_MONTHS = 60 * 60 * 24 * 30 * 6
+SIX_MONTHS = 60 * 60 * 24 * 30 * 6  # noqa: WPS432
 
-class DBEngines(str, Enum):
+
+class DBEngines(str, Enum):  # noqa: WPS600
     """Constraints for valid engine types that are supported by this component."""
     postgres = 'postgresql'
     mariadb = 'mysql'
     mysql = 'mysql'
     mysql_rds = 'mysql_rds'
     mongodb = 'mongodb'
 
+
 class OLVaultDatabaseConfig(BaseModel):
     db_name: Text
     mount_point: Text
     db_admin_username: Text
-    db_admin_password: SecretStr
+    db_admin_password: Text
     verify_connection: bool = True
-    db_host: Text
+    db_host: Union[Text, Output[Text]]
     max_ttl: int = SIX_MONTHS
     default_ttl: int = SIX_MONTHS
 
+    class Config:  # noqa: WPS431, D106
+        arbitrary_types_allowed = True
+
 
 class OLVaultPostgresDatabaseConfig(OLVaultDatabaseConfig):
-    db_port: Int = DEFAULT_PORT_POSTGRES
-    db_connection: Text = 'postgresql://{{db_user}}:{{db_pass}}@{{db_host}}:{{db_port}}/{{db_name}}'
+    db_port: int = DEFAULT_PORT_POSTGRES
+    # The db_connection strings are passed through the `.format` method so the variables that need to remain in the
+    # template to be passed to Vault are wrapped in 4 pairs of braces. TMM 2020-09-01
+    db_connection: Text = 'postgresql://{{{{username}}}}:{{{{password}}}}@{db_host}:{db_port}/{db_name}'
     db_type: Text = DBEngines.postgres.value
-    role_statements: Dict[Text, Text] = postgresql_sql_statements
+    role_statements: Dict[Text, Dict[Text, Text]] = postgres_sql_statements
 
 
 class OLVaultMysqlDatabaseConfig(OLVaultDatabaseConfig):
-    db_port: Int = DEFAULT_PORT_MYSQL
-    db_connection: Text = '{{db_user}}:{{db_pass}}@tcp({{db_host}}:{{db_port}})/'
+    db_port: int = DEFAULT_PORT_MYSQL
+    db_connection: Text = '{{{{username}}}}:{{{{password}}}}@tcp({db_host}:{db_port})/'
```

Comment:
Where are the `username` and `password` defined?

Reply:
Those are getting templated in from the `data` block that gets passed to the database secret backend class. It also allows for rotating the root credentials within Vault so that there's no way to recover them exernally.